### ISSUE TITLE
Only recompile changed files?

### ISSUE THREAD

**Issue by sid-kap** [tpolecat/tut#64]

**Comment by tpolecat** [comment]
There is some subtlety here because you want to recompile tut files if they have changed *or* if source code they depend on has changed. Otherwise if you change your code and wish to recompile your tut files you have to `touch` them or `clean` or something. This behavior existed for a while in an earlier version and I removed it because it was too confusing.

I'll leave this open because it's a good idea, I just don't know how to implement it. I suppose there could be a `tutChangedOnly` command or something. Any ideas?

**Comment by sid-kap** [comment]
Honestly, for my use case, I would be totally fine with recompiling only the files that have changed -- I do a clean rebuild when publishing anyway. But I agree that it would be nice to have correctness.

Here's an idea (not sure if this is possible):
We could convert each source file to a scala file that contains just the code snippets from the file. Then we could have sbt run a compile using the normal incremental compile system. Then we could look at which sources were recompiled to figure out which sources we need to re-run.

I would hope that there are SBT functions we can call to get information about dependencies and such. I'm not sure how that API works though. It's probably not a good idea to use last-modified times on files for something like this.

**Comment by tpolecat** [comment]
Yeah I thought about generating scala modules for each page and then using the built-in dependency checker to figure out which need recompiling, but I don't know how to do that. I'll dig around and see what I can find. Will probably end up asking @jsuereth for help. That's what usually happens :stuck_out_tongue_winking_eye:.

**Comment by tpolecat** [comment]
@noelwelsh re: #109 do you know about `tutOnly <filename>`? It just compiles the file you ask for, and you can use `~` with it.

But I think it might be sensible to say ok look I'm only writing doc, the source isn't changing, so just look at file timestamps in the doc.

**Comment by noelwelsh** [comment]
Ah, that new to me. I think it will be very useful. Thanks!

**Comment by tpolecat** [comment]
(@noelwelsh also, you can use tab completion with `tutOnly`)
